crates/ruff/src/noqa.rs

```diff
@@ -38,32 +40,93 @@ pub(crate) enum Directive<'a> {
+            // to ignore all rules.
+            let noqa_literal_end = mat.end();
+            return Some(if text[noqa_literal_end..].starts_with(':') {
+                // E.g., `# noqa: F401, F841`.
+                let mut codes_start = noqa_literal_end;
+
+                // Skip the `:` character.
+                codes_start += 1;
+
+                // Skip any whitespace between the `:` and the codes.
+                codes_start += text[codes_start..]
+                    .find(|c: char| !c.is_whitespace())
+                    .unwrap_or(0);
+
+                // Extract the comma-separated list of codes.
+                let mut codes = vec![];
+                let mut codes_end = codes_start;
+                let mut leading_space = 0;
+                while let Some(code) = Directive::lex_code(&text[codes_end + leading_space..]) {
+                    codes.push(code);
+                    codes_end += leading_space;
+                    codes_end += code.len();
+
+                    // Codes can be comma- or whitespace-delimited. Compute the length of the
+                    // delimiter, but only add it in the next iteration, once we find the next
+                    // code.
+                    leading_space = text[codes_end..]
+                        .find(|c: char| !(c.is_whitespace() || c == ','))
+                        .unwrap_or(0);
+                    if leading_space == 0 {
+                        break;
+                    }
                 }
-                let range = TextRange::at(
-                    TextSize::try_from(noqa.start()).unwrap().add(offset),
-                    noqa.as_str().text_len(),
+
+                let range = TextRange::new(
+                    TextSize::try_from(comment_start).unwrap(),
+                    TextSize::try_from(codes_end).unwrap(),
                 );
-                Some(Self::Codes(Codes { range, codes }))
-            }
-            (Some(noqa), None) => {
-                let range = TextRange::at(
-                    TextSize::try_from(noqa.start()).unwrap().add(offset),
-                    noqa.as_str().text_len(),
+
+                Self::Codes(Codes {
+                    range: range.add(offset),
+                    codes,
+                })
+            } else {
+                // E.g., `# noqa`.
+                let range = TextRange::new(
+                    TextSize::try_from(comment_start).unwrap(),
+                    TextSize::try_from(noqa_literal_end).unwrap(),
                 );
-                Some(Self::All(All { range }))
-            }
-            _ => None,
+                Self::All(All {
+                    range: range.add(offset),
+                })
+            });
+        }
+
+        None
+    }
+
+    /// Lex an individual rule code (e.g., `F401`).
+    fn lex_code(text: &str) -> Option<&str> {
+        // Extract, e.g., the `F` in `F401`.
+        let prefix = text.chars().take_while(char::is_ascii_uppercase).count();
```

Comment:
what would this code do if i did `# noqa: f401, f402`, i.e. would it silently do the wrong thing?

Reply:
It'll treat that as if it's a code-specific noqa with no matching codes -- which I think is correct. We're going to warn on that in a subsequent PR. (Today, it would be even worse: that would be treated as a blanket noqa, equivalent to `# noqa`, to ignore _all_ errors.)